src/ucs/datastruct/callbackq.c

```diff
@@ -247,3 +249,70 @@ ucs_status_t ucs_callbackq_remove_safe(ucs_callbackq_t *cbq, ucs_callback_t cb,
     ucs_callbackq_leave(cbq);
     return UCS_OK;
 }
+
+/* called with lock held */
+static ucs_callbackq_slow_elem_t* UCS_F_MAYBE_UNUSED
+ucs_callbackq_slow_path_find(ucs_callbackq_t *cbq,
+                             ucs_callbackq_slow_elem_t *elem)
+{
+    ucs_callbackq_slow_elem_t *it_elem;
+
+    ucs_queue_for_each(it_elem, &cbq->slow_path, queue){
+        if (it_elem == elem) {
+            return elem;
+        }
+    }
+    return NULL;
+}
+
+static void ucs_callbackq_slow_path_cb(void *arg)
+{
+    ucs_callbackq_t *cbq = arg;
+    ucs_callbackq_slow_elem_t *elem;
+    ucs_queue_iter_t iter;
+
+    ucs_callbackq_enter(cbq);
+
+    /* TODO avoid locks on every iteration. For instance implement
+     * lazy remove with additional list of elements to be removed. */
+    ucs_queue_for_each_safe(elem, iter, &cbq->slow_path, queue) {
+        ucs_callbackq_leave(cbq);
+        elem->cb(elem);
+        ucs_callbackq_enter(cbq);
+     }
+
+    ucs_callbackq_leave(cbq);
+}
+
+void ucs_callbackq_add_slow_path(ucs_callbackq_t *cbq,
+                                 ucs_callbackq_slow_elem_t* elem)
+{
+    ucs_status_t status;
+
+    ucs_callbackq_enter(cbq);
+
+    /* Multiple calls with the same element are not allowed */
+    ucs_assert(ucs_callbackq_slow_path_find(cbq, elem) == NULL);
```

Comment:
now thinking of this, it could *really* slow down the debug version ( create O(n^2) ), so we should remove the check unfortunately